### PR TITLE
Update Create Site URL inside Calypso for existing users

### PR DESCRIPTION
**Pull request by razvanpapadopol**
This is a follow up of #46484 to improve UX by skipping a redirect. Current solution has been first mentioned in pbxNRc-qE-p2#comment-1048

This PR is a dependency for the existing users non-EN experiment (pbxNRc-vn-p2) in context of a current exPlat limitation to fetch experiments outside a React application (tracked in 536-gh-Automattic/experimentation-platform).

#### Changes proposed in this Pull Request

* Update URL for all the "Create Site" buttons displayed in Calypso for logged in users.
* Cleanup redirect logic from `/start` to `/new` in signup controller to prevent any unwanted redirect.

#### Other changes
* Add `new_onboarding_url` property to config files.
* Re-use `<NoSitesMessage />` component in `<BlogsSettings />`.
* Add `ref=calypso-inline-help` when creating a site from `InlineHelp` block.
* Add `ref=me-account-settings` query when creating a site from Account Settings page.
* Add `ref=me-account-close` query when creating a site from `AccountCloseConfirmDialog`.
* Add `ref=calypso-sidebar` query when creating a site from `MySitesSidebar`.

#### Testing instructions
* As an existing user with EN language set in your account and without any site, check if correct link to create a site (`/new?ref={REF_SLUG}` is available for "Create Site" buttons on:
  * `/stats/day`
  * `/me` and go to "Profile Links" section  https://cloudup.com/cKwt6mSB7HY
  * `/me` and click on the bottom-right question mark icon to open Inline Help and search for "Create" 
  * `/me/account` and look for the button below "Primary Site" label
  * `/me/notifications`
  * `/me/account/close` and look for the "Start a new site" buttonhttps://cloudup.com/cobxG1YYddv

* Once you create your first site, go to My Home (`/home/{SITE_SLUG}`) and look for "Add New Site" button in the left sidebar.
* Once you create your second site, go to My Home (`/home/{SITE_SLUG}`), click on "Switch Site" button at the top of left sidebar and look for "Add New Site" button below the site entries.
* As an existing user with non-EN language set in your account, if you repeat all the above, you should get redirected to `/start?ref={REF_SLUG}` except if you are assigned to `treatment` variation of `new_onboarding_existing_users_non_en_v4` exPlat test (introduced in #48063 and described in pbxNRc-vn-p2)

### CODE REVIEW

**Comment by sarayourfriend** [comment]
@razvanpapadopol Was there anything in particular about this PR that you wanted Team Calypso input for?

**Comment by razvanpapadopol** [comment]
@saramarcondes I don't have something very specific in mind but it may be good for someone in Team Calypso to have a look and be aware of this change since it is touching a lot of places around Calypso.

Even if it feels like a light refactoring, it's changing the "Create Site" button URL from being a static string (coming from `config`) to a selector on Calypso store. I don't think this should have a significant performance impact but it's worth checking with you.

**Comment by yansern** [comment]
**TLDR version: All test instructions passes as described.**

## As EN User 
For `/stats/day`:

```html
<a class="button empty-content__action is-primary" href="/new?ref=calypso-nosites">Create Site</a>
```

For `/me` under **Profile Links > Add > Add Wordpress Site**:

```
Clicking on the button redirected to "/new?ref=me-profile-links".
```

For **Create A New Site** button under **Inline Help**:

```html
<div class="inline-help__results-cell"><a href="/new?ref=calypso-inline-help">...</a></div>
```

For `/me/account` under **Primary Site**:

```html
<fieldset class="form-fieldset"><label for="primary_site_ID" class="form-label">Primary Site</label><a href="/new?ref=me-account-settings" class="button">Add New Site</a></fieldset>
```

For `/me/notifications`:

```html
<a class="button empty-content__action is-primary" href="/new?ref=calypso-nosites">Create Site</a>
```

  For `/me/account/close` after clicking **Close My Account** button and the dialog pops up:

```html
<div class="account-close__confirm-dialog-alternative"><a href="/new?ref=me-account-close" class="button account-close__confirm-dialog-alternative-action" data-tracks-label="Start a new site">...</a></div>
```

After creating first site, the **Add New Site** button under **My Home > Customizer Sidebar**:

```html
<a class="sidebar__menu-link" href="/new?ref=calypso-sidebar">...</a>
```

 After creating second site, the **Add New Site** button under **My Home > Customizer Sidebar > Switch Site**:

```html
<a href="/new?ref=calypso-selector" class="button is-borderless">...</a>
```

## As Non-EN User

For `/stats/day`:

```html
<a class="button empty-content__action is-primary" href="/start?ref=calypso-nosites">Site aanmaken</a>
```

For `/me` under **Profile Links > Add > Add Wordpress Site**:

```
Clicking on the button redirected to "/start?ref=me-profile-links".
```

For **Create A New Site** button under **Inline Help**:

```html
<div class="inline-help__results-cell"><a href="/start?ref=calypso-inline-help">...<span>Nieuwe site&nbsp;maken</span></a></div>
```

For `/me/account` under **Primary Site**:

```html
<fieldset class="form-fieldset"><label for="primary_site_ID" class="form-label">Primaire site</label><a href="/start?ref=me-account-settings" class="button">Nieuwe site toevoegen</a></fieldset>
```

For `/me/notifications`:

```html
<a class="button empty-content__action is-primary" href="/start?ref=calypso-nosites">Site aanmaken</a>
```

For `/me/account/close` after clicking **Close My Account** button and the dialog pops up:

```html
<a href="/start?ref=me-account-close" class="button account-close__confirm-dialog-alternative-action" data-tracks-label="Start a new site">Start een nieuwe site...</a>
```

After creating first site, the **Add New Site** button under **My Home > Customizer Sidebar**:

```html
<a class="sidebar__menu-link" href="/start?ref=calypso-sidebar">...<span ...>Nieuwe site toevoegen</span></a>
```

 After creating second site, the **Add New Site** button under **My Home > Customizer Sidebar > Switch Site**:

```html
<a href="/start?ref=calypso-selector" class="button is-borderless">...Nieuwe site toevoegen</a>
```

## As Non-EN User under "treatment"

For this, I manually modified `get-onboarding-url.js` setting `existingUsersOnboardingVariant` to `treatment`.

Non-EN user now uses `/new` instead of `/start` like in the EN User test group above.

**Review by yansern — APPROVED** [review]
Code LGTM.

In summary:
- All places that uses `config( 'signup_url' )` is now replaced with `getOnboardingUrl()`.
- `getOnboadingUrl()` from `calypso/state/selectors/get-onboarding-url` is exposed as prop to various components.
- EN-users get `/new`.
- Non-EN users get `/start`.
- Non-EN users under explat test get `/new`;

File: config/_shared.json [hunk 1]
```diff
@@ -122,6 +122,7 @@
 	"rtl": false,
 	"siftscience_key": false,
 	"signup_url": false,
+	"new_onboarding_url": false,
```
Comment (tyxla) [inline config/_shared.json]: If I may nitpick on the config key naming here: `new` can lose its meaning outside of the context of this present project, especially for someone who hasn't worked on it that's looking at the code in half a year from now. Is it possible to use terms that make sense and bring additional context?

File: client/state/selectors/get-onboarding-url.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * Returns the onboarding URL.
+ *
+ *
+ * @param {object}  state Global state tree
+ * @returns {string}  URL of the onboarding flow for existing users.
+ */
+
+export default function getOnboardingUrl( state ) {
+	const userLocale = getCurrentUserLocale( state );
+	if ( [ 'en', 'en-gb' ].includes( userLocale ) ) {
```
Comment (tyxla) [inline client/state/selectors/get-onboarding-url.js]: Nit: can we define this array outside of the selector so it doesn't get redeclared on every selector call?

File: client/components/site-selector/add-site.jsx [hunk 3]
```diff
@@ -34,4 +34,9 @@ class SiteSelectorAddSite extends Component {
 	}
 }
 
-export default connect( null, { recordTracksEvent } )( localize( SiteSelectorAddSite ) );
+export default connect(
+	( state ) => ( {
+		onboardingUrl: getOnboardingUrl( state ),
```
Comment (tyxla) [inline client/components/site-selector/add-site.jsx:39]: The site selector is used in Jetpack Cloud, too. I can see we haven't taken the Jetpack Cloud environments into account - has this been considered? Will this change have any unexpected effects there? 

cc @Automattic/jetpack-voyager for their Jetpack Cloud expertise.

File: client/my-sites/controller.js [hunk 4]
```diff
@@ -137,7 +137,7 @@ function renderNoVisibleSites( context ) {
 		action: i18n.translate( 'Change Visibility' ),
 		actionURL: '//dashboard.wordpress.com/wp-admin/index.php?page=my-blogs',
 		secondaryAction: i18n.translate( 'Create New Site' ),
-		secondaryActionURL: `${ signup_url }?ref=calypso-nosites`,
+		secondaryActionURL: `${ getOnboardingUrl() }?ref=calypso-nosites`,
```
Comment (tyxla) [inline client/my-sites/controller.js]: Did we mean to call this without a `state` arg?

File: client/me/profile-links-add-wordpress/index.jsx [hunk 5]
```diff
@@ -171,6 +171,7 @@ class ProfileLinksAddWordPress extends Component {
 						{
 							components: {
 								jetpackLink: (
+									/* eslint-disable-next-line jsx-a11y/anchor-is-valid */
```
Comment (tyxla) [inline client/me/profile-links-add-wordpress/index.jsx]: I'd suggest removing this. It's a valid ESLint warning that I don't see a good reason to suppress.

Comment (jsnajdr) [reply client/me/profile-links-add-wordpress/index.jsx]: The `a` element should probably be a `button`. Do we have a convenient way to style `button` so that it looks like an inline hyperlink?

Alternatively, it could remain an `a` element, but specify `href="http://jetpack.me"`, with a `_target` attribute to open a new window. And the `a` element can still have an `onClick` handler that records the tracks events and calls `this.onCancel()`.

File: config/_shared.json (hunk 1 above)
Comment (razvanpapadopol) [reply config/_shared.json]: This seems to be the official name ( pcr1an-2g-p2 ) but I agree that it makes sense to replace it here with the former name of the flow which is mostly used in development (including in the project files [path](https://github.com/Automattic/wp-calypso/tree/trunk/client/landing/gutenboarding)).

**Review by jessie-ross — COMMENTED** [review]
Just going off [the little checklist we've just put up for Calypso](https://github.com/Automattic/experimentation-platform/wiki/Migration-guide-from-%60calypso-abtest%60#implementation-checklist):

- [x] Is your experiment assignment being fetched?


   It looks like you need to fetch your experiment by adding in:
   [```<Experiment name="new_onboarding_existing_users_non_en_v4" />```](https://github.com/Automattic/experimentation-platform/wiki/Migration-guide-from-%60calypso-abtest%60#accessing-explat-state-directly) 

- [x] Are you branching the code on `'treatment'`?
- [x] Will `variation === null` or `variation === 'control'` both cause the default experience?
- [ ] Have you accounted for loading state?
   
   This depends on where you put the `<Experiment` component mentioned above: will it be early enough to be sure the experiment will already be loaded? To me that seems necessary here but in case you think otherwise you would also add a loading state (in this case it would be like a disabled state on the button). 

Otherwise looking good 🙂

File: client/me/profile-links-add-wordpress/index.jsx (hunk 5 above)
Comment (razvanpapadopol) [reply client/me/profile-links-add-wordpress/index.jsx]: Thanks for the suggestion, @jsnajdr
Fixed in https://github.com/Automattic/wp-calypso/pull/48350/commits/6fc709cf058c649ac94d00937553e6d8e9675d45

File: client/my-sites/controller.js (hunk 4 above)
Comment (razvanpapadopol) [reply client/my-sites/controller.js]: Nop, my bad 🤦  
Fixed in https://github.com/Automattic/wp-calypso/pull/48350/commits/ee9761c3130cce82926a739153bc4495a50eff90

File: client/me/profile-links-add-wordpress/index.jsx (hunk 5 above)
Comment (sarayourfriend) [reply client/me/profile-links-add-wordpress/index.jsx]: FWIW: `Button` from `@automattic/components` has a `plain` which will reset all styles, making it easy to style it as an anchor. But probably better is to use `@wordpress/components` `Button` which has a `isLink` prop.

Comment (razvanpapadopol) [reply client/me/profile-links-add-wordpress/index.jsx]: > But probably better is to use @wordpress/components Button which has a isLink prop.

I tried that but the styling doesn't seem right.
Screenshot 2020-12-16 at 17 50 54
I'll leave the `a` element with `href` and `onClick` attributes which seems to be the cleanest solution that also makes eslint happy.

File: config/_shared.json (hunk 1 above)
Comment (jsnajdr) [reply config/_shared.json]: The "Gutenboarding" name is used very consistently across the codebase and is much more unique and descriptive than just "new". This name is not intended to be ever seen by the user in the UI, but for naming variables or feature flags, it's the right name to use. Here, `gutenboarding_url` is guaranteed to be understood even after many months and years.

File: client/components/site-selector/add-site.jsx (hunk 3 above)
Comment (tyxla) [reply client/components/site-selector/add-site.jsx:39]: FWIW I've pinged the Voyager team in Slack about this: p1608201558180500-slack-CQXNTE9K9

File: config/_shared.json (hunk 1 above)
Comment (tyxla) [reply config/_shared.json]: `gutenboarding_url` makes so much more sense indeed 👍  Thanks for considering it @razvanpapadopol

File: client/me/profile-links-add-wordpress/index.jsx [hunk 6]
```diff
@@ -172,7 +170,9 @@ class ProfileLinksAddWordPress extends Component {
 							components: {
 								jetpackLink: (
 									<a
-										href="#"
+										href="http://jetpack.me"
```
Comment (tyxla) [inline client/me/profile-links-add-wordpress/index.jsx]: I know this was like this before, but since we're touching it already, would you mind making it `https`?

File: client/my-sites/controller.js [hunk 7]
```diff
@@ -112,7 +113,7 @@ function renderNoVisibleSites( context ) {
 	const { getState } = getStore( context );
 	const currentUser = getCurrentUser( getState() );
 	const hiddenSites = currentUser && currentUser.site_count - currentUser.visible_site_count;
-	const signup_url = config( 'signup_url' );
+	const onboardingUrl = getOnboardingUrl( getState() );
```
Comment (tyxla) [inline client/my-sites/controller.js]: Nit: Could we call `getState()` just once in this function? Perhaps if we store it in a var above?

**Review by tyxla — COMMENTED** [review]
I haven't tested, but have no other concerns about the code, other than the nits I've already posted. 

The bundle size changes are expected IMHO, and I don't see anything concerning there.

Please let me know if you have any questions or need anything specific reviewed.

File: client/components/site-selector/add-site.jsx (hunk 3 above)
Comment (rcanepa) [reply client/components/site-selector/add-site.jsx:39]: Thanks for bringing this up Marin. I can see two broken flows in Jetpack Cloud:

1) User with no sites.
CloudNoSites

2) User with sites wants to add a new site.
CloudSwitchSite

However, these two issues were already present in production and were not introduced by these changes. I think the sensible thing to do is to create a card in our Maintainance board and fix them in a follow-up PR.

Would you all be okay with that?

Comment (tyxla) [reply client/components/site-selector/add-site.jsx:39]: Thanks for chiming in so quickly, @rcanepa. 

As far as this PR is concerned, it seems like there is no regression introduced on the Jetpack Cloud side, so it should be good to go from this side @razvanpapadopol.

And I appreciate you taking the time to report and fix this issue separately. Cross-team collab FTW!
